Fix incorrect array initialization with string literal (fixes #112189)

awson opened the pull request:
Following [the suggestion of @zygoloid](https://github.com/llvm/llvm-project/issues/112189#issuecomment-3172788105) `updateStringLiteralType` now recursively clones subexpressions which types it modifies.

Done as a separate static function and not inlined directly into the `updateStringLiteralType` function for maintainability.

I probably should have done this for the dependent case only, but the function is buried deep in the call stack and I didn't bother.

Fixes #112189

shafik reviewed (commented):
please update the subject of the PR to better describe what the PR is actually fixing.

awson committed:
`updateStringLiteralType` recursively clones subexpressions which types it modifies.

Done as a separate static function and not inlined directly into the `updateStringLiteralType` function for maintainability.

cor3ntin commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -168,16 +168,95 @@ bool Sema::IsStringInit(Expr *Init, const ArrayType *AT) {
+      std::copy(GSEAEs.begin(), GSEAEs.end(), NewGSEAEs);
+      NewGSEAEs[GSE->getResultIndex()] = CloneDrilled(GSE->getResultExpr(), C);
+
+#define CREATE_GSE(ctrl)                                                       \
+  GenericSelectionExpr::Create(                                                \
+      C, GSE->getGenericLoc(), ctrl, GSE->getAssocTypeSourceInfos(),           \
+      ArrayRef<Expr *>(NewGSEAEs, GSEAEs.size()), GSE->getDefaultLoc(),        \
+      GSE->getRParenLoc(), GSE->containsUnexpandedParameterPack(),             \
+      GSE->getResultIndex())
+
+      return GSE->isExprPredicate() ? CREATE_GSE(GSE->getControllingExpr())
+                                    : CREATE_GSE(GSE->getControllingType());
+#undef CREATE_GSE
```
Lets use a lambda or a function here

cor3ntin commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -168,16 +168,95 @@ bool Sema::IsStringInit(Expr *Init, const ArrayType *AT) {
+// Exactly follow `IgnoreParensSingleStep` (`AST/IgnoreExpr.h`)
+// We only clone those subexpressions which `IgnoreParensSingleStep` drills down
+// to.
+static Expr *CloneDrilled(Expr *E, ASTContext &C) {
```
Can we find a better name? `CloneEnclosedStringLiteral` or smth?

awson replied:
Changed to `CloneWithRecurseToInnermostSL`. I thing it much better reflects what is going on here.

cor3ntin commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -168,16 +168,95 @@ bool Sema::IsStringInit(Expr *Init, const ArrayType *AT) {
 /// to match the type of the object which it is initializing.
-static void updateStringLiteralType(Expr *E, QualType Ty) {
+static Expr *updateStringLiteralType(Expr *E, QualType Ty, Sema &S) {
+  Expr *ENew = CloneDrilled(E, S.Context);
+  E = ENew;
```
```suggestion
  E = CloneDrilled(E, S.Context);
```

cor3ntin commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -168,16 +168,95 @@ bool Sema::IsStringInit(Expr *Init, const ArrayType *AT) {
       break;
     E = IgnoreParensSingleStep(E);
   }
+  return ENew;
```
That should be E, right?

awson replied:
Nope.

We need to keep `ENew` for return.

`E` drills down to innermost string literal.

cor3ntin commented:
Thanks for working on this. 

I have some concerns that we do that transformation even in non template contexts, it seems a fairly big hammer. Can you try to find ways to limit how often the cloning is done? id you explore the other suggestion made by @zygoloid at all ? (to do it in TreeTransform)
